design/volume-group-snapshot.md

```diff
@@ -0,0 +1,428 @@
+        - In all the instances, the VGS label key will be present on the backup spec, this makes the label key accessible to plugins during the execution of backup operation.
+    - This label will enable velero to filter the PVC to be included in the VGS spec.
+    - Users will have to label the PVCs before invoking the backup operation.
+    - This label would act as a group identifier for the PVCs to be grouped under a specific VGS.
+    - It will be used to collect the PVCs to be used for a particular instance of VGS object.
+
+- Changes to the Existing PVC ItemBlockAction plugin:
+    - Currently the PVC IBA plugin is applied to PVCs and adds the RelatedItems for the particular PVC into the ItemBlock.
+    - At first it checks whether the PVC is bound and VolumeName is non-empty.
+    - Then it adds the related PV under the list of relatedItems.
+    - Following on, the plugin adds the pods mounting the PVC as relatedItems.
+    - Now we need to extend this PVC IBA plugin to add the PVCs to be grouped for a particular VGS object, so that they are processed together under an ItemBlock by Velero.
+        - First we will check if the PVC that is being processed by the plugin has the user specified VGS label.
+        - If it is present then we will execute a List call in the namespace with the label as a matching criteria and see if this results in any PVCs (other than the current one).
+        - If there are PVCs matching the criteria then we add the PVCs to the relatedItems list.
+        - This helps in building the ItemBlock we need for VGS processing, i.e. we have the relevant pods and PVCs in the ItemBlock.
+
+- Updates to CSI PVC plugin:
+    - This is the plugin which creates VolumeSnapshots for the CSI PVCs if required.
+    - We will have to update this plugin to:
+        - Create VGS for the group of PVCs
+            - At First we will check if PVC has the user specified VGS label.
+            - If yes, then it implies it needs to be part of a VGS instance.
+            - Check if VGS exists for this PVC or not, we don't want to recreate VGS object.
+            - We can check if VGS exists by listing the VGS in the namespace using the backup UUID alongwith the label used for the VGS (we will be creating VGS with backup UUID and user specified VGS  group label) and see if the PVC is part of any of the existing VGS for the particular backup operation we are currently in.
+            - If not then create the VGS object with the required labels (in this case backup UUID and the user specified VGS group label) and also add the PVC names (that are included in the VGS) as annotations. (This will be useful for our VGS restore operation where we would need PVC resource identifiers).
+            - Once VGS is created, add the VGS as an additional item to be included in the backup.
+        - Skip creation of VS for PVC that are already included in VGS
```

Comment:
We expect the label specified to the PVC is added before the backup.
If it is added/modified during the backup, there will be a chaos, e.g., at T0, when processing PVC1, a VGS is created and include PVC2, because it HAS the label. Then at T1, when processing PVC2, the label is removed, so a single VS is also crated for PVC2.

Therefore, we should have a mechanism to detect this case. Once we detect it, we can bypass the creation of VS for PVC2 or error out.

Reply:
hmm so in order to handle this case should we annotate the backup object with an initial record/snapshot of grouping ? This can be done during the PVC IBA plugin and then we use this record for further processing in CSI PVC plugin.

Reply:
I don't think it is a good idea to modify the source object (e.g., PVC here) during the backup for any purpose.

I think we can collect the PVCs with the VGS label at the collection stage and refer to the record instead of querying it from API server.
But there may be a performance issue, when there is a large number of PVCs in the backup.

Reply:
@Lyndon-Li by backup object I meant the Velero Backup CR and not the PVC object.

Reply:
Yeah, I don't think it really makes sense to add to the collection stage here. That also makes the workflow more convoluted. The ItemBlock and ItemBlockAction plugin APIs already allow for grouping things. I think "user adds new labels while backup happens" is an edge case that we can just document. Users messing with item metadata during backup unrelated to this can have similar unpredictable effects (i.e. adding/removing labels when we're using labels for volume policies or item selection, etc. Existing plugins already query live objects sometimes -- so there's nothing new here.

Reply:
> by backup object I meant the Velero Backup CR and not the PVC object.
This doesn't look quit ideal either because it may result in many annotations if there are many PVCs and VGS.

Reply:
I am fine to document it, if we don't have a good solution.
Then let's clarify this in the design and then add the same in the doc.

Reply:
Done, updated the PR, added a note.